plan/newplanbuilder.go

```diff
@@ -479,61 +477,226 @@ func (b *planBuilder) extractAggFunc(sel *ast.SelectStmt) (
 	return aggList, havingMapper, orderByMapper, totalAggMapper
 }
 
+// havingAndOrderbyResolver resolves having/orderby's ast expression to expression.Expression
+type havingAndOrderbyResolver struct {
+	proj    *Projection
+	inExpr  bool
+	orderBy bool
+	err     error
+	mapper  map[*ast.ColumnNameExpr]expression.Expression
+}
+
+func (e *havingAndOrderbyResolver) addProjectionExpr(v *ast.ColumnNameExpr, projCol *expression.Column) {
+	e.proj.Exprs = append(e.proj.Exprs, projCol)
+	schemaCols, _ := projCol.DeepCopy().(*expression.Column)
+	e.mapper[v] = schemaCols
+	e.proj.schema = append(e.proj.schema, schemaCols)
+}
+
+func (e *havingAndOrderbyResolver) Enter(inNode ast.Node) (ast.Node, bool) {
+	switch v := inNode.(type) {
+	case *ast.ValueExpr, *ast.ColumnName, *ast.ParenthesesExpr:
+	case *ast.ColumnNameExpr:
+		var first, second expression.Schema
+		var col *expression.Column
+		fromSchema := e.proj.GetChildByIndex(0).GetSchema()
+		fromSchemaFirst := e.orderBy && e.inExpr
+		if fromSchemaFirst {
+			first, second = fromSchema, e.proj.GetSchema()
+			col, e.err = first.FindColumn(v.Name)
+		} else {
+			first, second = e.proj.GetSchema(), fromSchema
+			col, e.err = first.FindSelectFieldColumn(v.Name, e.proj.Exprs)
+		}
+		if e.err != nil {
+			return inNode, true
+		}
+		if col == nil {
+			if fromSchemaFirst {
+				col, e.err = second.FindSelectFieldColumn(v.Name, e.proj.Exprs)
+			} else {
+				col, e.err = second.FindColumn(v.Name)
+			}
+
+			if e.err != nil {
+				return inNode, true
+			}
+			if col == nil {
+				e.err = errors.Errorf("Can't find Column %s", v.Name.Name)
+				return inNode, true
+			}
+			if !fromSchemaFirst {
+				e.addProjectionExpr(v, col)
+			} else {
+				e.mapper[v] = col
+			}
+		} else if fromSchemaFirst {
+			e.addProjectionExpr(v, col)
+		} else {
+			e.mapper[v] = col
+		}
+	case *ast.SubqueryExpr, *ast.CompareSubqueryExpr, *ast.ExistsSubqueryExpr:
+		return inNode, true
+	default:
+		e.inExpr = true
+	}
+	return inNode, false
+}
+
+func (e *havingAndOrderbyResolver) Leave(inNode ast.Node) (ast.Node, bool) {
+	return inNode, true
+}
+
+type gbyResolver struct {
+	fields []*ast.SelectField
+	schema expression.Schema
+	err    error
+	inExpr bool
+}
+
+func (g *gbyResolver) Enter(inNode ast.Node) (ast.Node, bool) {
+	switch inNode.(type) {
+	case *ast.SubqueryExpr, *ast.CompareSubqueryExpr, *ast.ExistsSubqueryExpr:
+		return inNode, true
+	case *ast.ValueExpr, *ast.ColumnNameExpr, *ast.ParenthesesExpr, *ast.ColumnName:
+	default:
+		g.inExpr = true
+	}
+	return inNode, false
+}
+
+func (g *gbyResolver) match(a *ast.ColumnName, b *ast.ColumnName) bool {
+	if a.Schema.L == "" || a.Schema.L == b.Schema.L {
+		if a.Table.L == "" || a.Table.L == b.Table.L {
+			return a.Name.L == b.Name.L
+		}
+	}
+	return false
+}
+
+func (g *gbyResolver) Leave(inNode ast.Node) (ast.Node, bool) {
+	switch v := inNode.(type) {
+	case *ast.ColumnNameExpr:
```

Comment:
Is there any other type you should handle? If not, we do not need the switch statement.